lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -258,18 +442,29 @@ def _call(self, method: Callable, json_data: JSON):
             es_request = self.assemble(json_data, context)
             path = es_request.get("path")
             url = urljoin(self.es_url, path)
-            self.logger.debug("ASSEMBLE returned URL {}", url)
+            self.logger.info(
+                "ASSEMBLE returned URL {!r}, {!r}", url, es_request.get("json")
+            )
+        except NoSelectedDatasets as e:
+            # The query builder reports that no datasets can be found; skip
+            # the Elasticsearch query and tell postprocess to build an empty
+            # client response as appropriate.
+            self.logger.info("{} query builder reports no matches: {}", klasname, e)
+            context["NODATA"] = True
         except Exception as e:
             self.logger.exception("{} assembly failed: {}", klasname, e)
             abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
 
         try:
-            # perform the Elasticsearch query
-            es_response = method(url, **es_request["kwargs"])
-            self.logger.debug(
-                "ES query response {}:{}", es_response.reason, es_response.status_code
-            )
-            es_response.raise_for_status()
+            if not context.get("NODATA", False):
+                # perform the Elasticsearch query
+                es_response = method(url, **es_request["kwargs"])
```

Comment:
Well, it's a hole in the Flask-level testing. I covered all the cases in the `test_query_builder.py` unit test, but this is proof that's not quite enough since it can overflow into here. Ugh.

I'll probably close this by pulling out the `es_response.json()` here into a variable which is otherwise `None` (or maybe `{}`), which the logic outside the `try` block can use safely.

And I suppose I need to add a Flask test case (as messy as they are) to cover the `NODATA` path. ☹️